src/permutations.jl

```diff
@@ -10,66 +10,122 @@ export
 
 
 struct Permutations{T}
-    a::T
-    t::Int
+    data::T
+    length::Int
 end
 
-Base.eltype(::Type{Permutations{T}}) where {T} = Vector{eltype(T)}
+function has_repeats(state::Vector{Int})
+    # This can be safely marked inbounds because of the type restriction in the signature.
+    # If the type restriction is ever loosened, please check safety of the `@inbounds`
+    @inbounds for outer in eachindex(state)
+        @inbounds for inner in (outer+1):lastindex(state)
+            if state[outer] == state[inner]
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function increment!(state::Vector{Int}, min::Int, max::Int)
+    state[end] += 1
+    for i in reverse(eachindex(state))[firstindex(state):end-1]
```

Comment:
```suggestion
    for i in lastindex(state):-1:(firstindex(state) - 1)
```
Subjective but I find this clearer

Reply:
Huh, I thought this was equivalent to what you had but I think this broke the tests.

Reply:
I read your suggestion and thought it was equivalent as well! Testing for the case `permutations(1:3, 2)`, we get that `state = fill(firstindex(p.data), p.length)` which is `[1, 1]`. Then 

- `collect(reverse(eachindex(state))[firstindex(state):end-1])` returns `[2]`
- `collect(lastindex(state):-1:(firstindex(state) - 1))` returns `[2, 1, 0]`